### PR TITLE
Update sort node to use alias reference when it will resolve missing columns

### PR DESCRIPTION
**Pull request by fulghum**
Fixes: https://github.com/dolthub/dolt/issues/3016

Other changes: 
- Refactors the existing OrderBy/GroupBy tests into a ScriptTest. 
- Introduces a new interface, `sql.Projector`, that unites `GroupBy`, `Window`, and `Project` if a caller just needs to get the projected expressions.

### CODE REVIEW

File: sql/analyzer/aliases.go [hunk 1]
```diff
@@ -149,6 +149,28 @@ func getTableAliases(n sql.Node, scope *Scope) (TableAliases, error) {
...
 
+// aliasedExpressionsInNode returns a map of the aliased expressions defined in the specified node, mapped from the
+// expression string to the alias name.
+func aliasedExpressionsInNode(n sql.Node) map[string]string {
+	var exprs []sql.Expression
+	switch n := n.(type) {
+	case sql.Projector:
+		exprs = n.ProjectedExprs()
+	case sql.Expressioner:
+		exprs = n.Expressions()
```
Comment (max-hoffman) [inline sql/analyzer/aliases.go:160]: nit, defaults for every switch
```suggestion
		exprs = n.Expressions()
	default:
```

File: sql/analyzer/resolve_orderby.go [hunk 2]
```diff
@@ -75,8 +78,14 @@ func pushdownSort(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel R
 			return n, transform.SameTree, nil
 		}
 
-		// If there are no columns required by the order by available, then move the order by
-		// below its child.
+		// If all the missing expressions are aliased, swap in the alias name and don't move the sort node
+		expressionToAliasName := aliasedExpressionsInNode(sort.Child)
```
Comment (max-hoffman) [inline sql/analyzer/resolve_orderby.go:82]: can a HAVING, FILTER or LIMIT node sit between the SORT and Projector? i.e. do we need to inspect the scope bottom-up? or top down iterate until we find nil or a Projector?

Comment (fulghum) [reply sql/analyzer/resolve_orderby.go:82]: Great question. Seems like `LIMIT` should always be above the `SORT` node to get the right results, but you have a great point about `FILTER` and `HAVING`. I just added some more tests to trigger this case, fixed the code, and pushed a new version. Thanks for suggesting this. 🙏

File: sql/core.go [hunk 3]
```diff
@@ -288,6 +288,14 @@ type OpaqueNode interface {
 	Opaque() bool
 }
 
+// Projector is a node that projects expressions for parent nodes to consume (i.e. GroupBy, Window, Project).
+type Projector interface {
+	// ProjectedExprs returns the list of expressions projected by this node.
+	ProjectedExprs() []Expression
+	// WithProjectedExprs returns a new Projector instance with the specified expressions set as its projected expressions.
```
Comment (max-hoffman) [inline sql/core.go:295]: nit: the docstring says we return a Projector, but the interface is Node

**Review by max-hoffman — APPROVED** [review]
LGTM!

File: sql/core.go (hunk 3 above)
Comment (fulghum) [reply sql/core.go:295]: good catch; fixed to return `sql.Projector`

**Pull request merged**